### PR TITLE
Source map support

### PR DESCRIPTION
**Pull request by DylanC**
After we merge #225 it would be nice to add support for source map usage. I'm not sure what is the best approach but I was thinking of continuing the work from #185. This branch includes the previous commits of the original PR. (Rebased on master)

### CODE REVIEW

**Comment by DylanC** [comment]
@davidmarkclements or @mcollina - Do you have any input as to what else is needed here?

File: lib/ticks-to-tree.js [hunk 1]
```diff
@@ -15,19 +19,22 @@ function hasFileLocation (jsFrameName) {
...
-    value: 0,
-    top: 0,
-    children: []
-  }
-  const unmerged = {
-    name: 'all stacks',
-    value: 0,
-    top: 0,
-    children: []
+async function ticksToTree (ticks, options = {}) {
```
Comment (mcollina) [inline lib/ticks-to-tree.js]: Is it necessary to turn this into an `async` function?

File: lib/ticks-to-tree.js [hunk 2]
```diff
@@ -99,71 +112,184 @@ function ticksToTree (ticks, options = {}) {
...
-
-      if (ix === stack.length - 1) frame.top++
-      if (ix === 0) merged.value += 1
-      frame.value++
-
-      lastFrame = frame
+  if (sourceMaps) {
+    await Promise.all(Object.keys(jsLibraries).map(async (jsLibrary) => {
+      return loadSourceMap(jsLibrary, sourceMaps)
+    }))
```
Comment (mcollina) [inline lib/ticks-to-tree.js]: This is going to be pretty expensive, this should use p-limit or p-map.

File: lib/ticks-to-tree.js [hunk 3]
```diff
@@ -99,71 +112,184 @@ function ticksToTree (ticks, options = {}) {
...
+  Object.keys(loadedSourceMaps).forEach((file) => {
+    if (loadedSourceMaps[file] !== false) {
+      loadedSourceMaps[file].destroy()
+    }
+  })
+
+  return trees
+}
+
+const loadedSourceMaps = {}
```
Comment (mcollina) [inline lib/ticks-to-tree.js]: This should not be a global. It creates a memory leak

File: lib/ticks-to-tree.js (hunk 1 above)
Comment (DylanC) [reply lib/ticks-to-tree.js]: Not sure. I'm picking up this from previous PR changes. I don't see why it is like this currently.

File: lib/ticks-to-tree.js (hunk 2 above)
Comment (DylanC) [reply lib/ticks-to-tree.js]: Will do!

File: lib/ticks-to-tree.js (hunk 3 above)
Comment (DylanC) [reply lib/ticks-to-tree.js]: No problem. I'll change this.

File: lib/ticks-to-tree.js (hunk 1 above)
Comment (DylanC) [reply lib/ticks-to-tree.js]: Ah, it uses await here:
`if (sourceMaps) {
    await Promise.all(Object.keys(jsLibraries).map(async (jsLibrary) => {
      return loadSourceMap(jsLibrary, sourceMaps)
    }))
  }`
I don't have an answer to your question really. I would likely need the original author to chime in. @pimlie ?

File: lib/ticks-to-tree.js [hunk 4]
```diff
@@ -99,71 +113,184 @@ function ticksToTree (ticks, options = {}) {
...
-
-      if (ix === stack.length - 1) frame.top++
-      if (ix === 0) merged.value += 1
-      frame.value++
-
-      lastFrame = frame
+  if (sourceMaps) {
+    await pMap(Object.keys(jsLibraries).map(async (jsLibrary) => {
+      return loadSourceMap(jsLibrary, sourceMaps)
+    }))
```
Comment (mcollina) [inline lib/ticks-to-tree.js]: Please add a concurrency of 16, check the p-map doc.

File: lib/ticks-to-tree.js [hunk 5]
```diff
@@ -99,71 +113,184 @@ function ticksToTree (ticks, options = {}) {
...
-      if (child === undefined) children.push(frame)
-      else frame = child
-
-      if (ix === stack.length - 1) frame.top++
-      if (ix === 0) merged.value += 1
-      frame.value++
-
-      lastFrame = frame
+  if (sourceMaps) {
+    await pMap(Object.keys(jsLibraries).map(async (jsLibrary) => {
```
Comment (mcollina) [inline lib/ticks-to-tree.js]: This is incorrect, the mapper is a the second argument. Check the p-map docs.

File: lib/ticks-to-tree.js (hunk 3 above)
Comment (DylanC) [reply lib/ticks-to-tree.js]: @mcollina Do you have any suggestions for this? loadedSourceMaps is used in 3 different functions.

File: lib/ticks-to-tree.js (hunk 5 above)
Comment (DylanC) [reply lib/ticks-to-tree.js]: I actually had it the other way around first but the linted complained that pMap wasn't being used.

File: lib/ticks-to-tree.js (hunk 3 above)
Comment (mcollina) [reply lib/ticks-to-tree.js]: You need to pass it as a parameter.

**Comment by DylanC** [comment]
@mcollina - I've changed loadedSourceMaps to not be global any longer. Anything else required here?

File: lib/ticks-to-tree.js [hunk 6]
```diff
@@ -99,71 +113,187 @@ function ticksToTree (ticks, options = {}) {
...
+    }
+
+    for (let i = 0; i < sourceMapFiles.length; i++) {
+      if (sourceMapFiles[i]) {
+        // convert relative source map paths to absolute paths
+        if (!isAbsolute(sourceMapFiles[i])) {
+          sourceMapFiles[i] = resolve(dirname(fileName), sourceMapFiles[i])
+        }
+
+        if (existsSync(sourceMapFiles[i])) {
```
Comment (mcollina) [inline lib/ticks-to-tree.js]: Please use an async version of this.

File: lib/ticks-to-tree.js [hunk 7]
```diff
@@ -99,71 +113,187 @@ function ticksToTree (ticks, options = {}) {
...
+          try {
+            debug(`loading source map from ${sourceMapFiles[i]}`)
+            const sourceMap = JSON.parse(readFileSync(sourceMapFiles[i]))
```
Comment (mcollina) [inline lib/ticks-to-tree.js]: please read the file asynchronously

File: lib/ticks-to-tree.js [hunk 8]
```diff
@@ -99,71 +113,187 @@ function ticksToTree (ticks, options = {}) {
...
+            loadedSourceMaps[fileName] = await new SourceMapConsumer(sourceMap)
+            if (!loadedSourceMaps[fileName].sourceRoot) {
+              loadedSourceMaps[fileName].sourceRoot = dirname(sourceMapFiles[i])
+            }
+            break
+          } catch (e) {
+            console.error(e)
```
Comment (mcollina) [inline lib/ticks-to-tree.js]: Are we sure this is correct? If we can't read the file, is it safe to just carry on?

File: lib/ticks-to-tree.js [hunk 9]
```diff
@@ -99,71 +113,187 @@ function ticksToTree (ticks, options = {}) {
...
-        else if (frame.S === 2) frame.name = '*' + frame.name
-        else if (frame.S === 3) frame.name = '~' + frame.name + ' [INLINABLE]'
-        else if (frame.S === 4) frame.name = '*' + frame.name + ' [INLINABLE]'
-        children.push(frame)
-      } else frame = child
+      if (source.source) {
+        name = `${key[0]} ${source.source}:${source.line}:${source.column}`
+      }
+    } catch (err) {
+      console.error(err, fileName, line, column)
```
Comment (mcollina) [inline lib/ticks-to-tree.js]: Are we sure this is correct? If we can't read the file, is it safe to just carry on?

File: lib/ticks-to-tree.js [hunk 10]
```diff
@@ -99,71 +113,187 @@ function ticksToTree (ticks, options = {}) {
...
+
+      // only rename when fileName is an absolute path
+      if (isAbsolute(fileName)) {
+        const relativeName = relative(options.relativePath, fileName)
+        frame.name = `${key[0]} ${relativeName}:${line}:${column}`
+      }
+    }
+  }
+  if (frame.children && frame.children.length) {
+    frame.children.forEach((child) => renameFrame(options, child, loadedSourceMaps))
```
Comment (mcollina) [inline lib/ticks-to-tree.js]: Please use a for loop, it'll be faster.

File: lib/ticks-to-tree.js [hunk 11]
```diff
@@ -99,71 +113,187 @@ function ticksToTree (ticks, options = {}) {
...
   }
+  delete frame.type
```
Comment (mcollina) [inline lib/ticks-to-tree.js]: can we just set type to undefined? It'll be faster.

File: lib/ticks-to-tree.js [hunk 12]
```diff
@@ -99,71 +113,187 @@ function ticksToTree (ticks, options = {}) {
...
+}
+
+function addToMergedTree (trees, stack) {
+  var lastFrame = null
+  stack.forEach((frame, ix) => {
```
Comment (mcollina) [inline lib/ticks-to-tree.js]: Please use a for(;;) loop.

File: lib/ticks-to-tree.js [hunk 13]
```diff
@@ -99,71 +113,187 @@ function ticksToTree (ticks, options = {}) {
...
+    if (ix === 0) trees.merged.value += 1
+    frame.value++
+
+    lastFrame = frame
+  })
+}
+
+function addToUnmergedTree (trees, stack) {
+  var lastFrame = null
+  stack.forEach((frame, ix) => {
```
Comment (mcollina) [inline lib/ticks-to-tree.js]: Please use a for(;;) loop

File: lib/ticks-to-tree.js (hunk 6 above)
Comment (DylanC) [reply lib/ticks-to-tree.js]: I was going to use fs.exists but the linter warned this is deprecated. Should I use fs.access instead?

Comment (DylanC) [reply lib/ticks-to-tree.js]: Using fs.access instead. :+1:

File: lib/ticks-to-tree.js [hunk 14]
```diff
@@ -99,71 +113,191 @@ function ticksToTree (ticks, options = {}) {
...
+  let loadedSourceMaps = {}
+  if (sourceMaps) {
+    const mapper = async (jsLibrary) => {
+      loadedSourceMaps = loadSourceMap(jsLibrary, sourceMaps, loadedSourceMaps)
+      return loadedSourceMaps
+    }
+    await pMap(Object.keys(jsLibraries), mapper, { concurrency: 16 })
+  }
+  if (sourceMaps || relativePath) {
+    Object.keys(trees).map((tree) => {
```
Comment (mcollina) [inline lib/ticks-to-tree.js]: this does not need to be a map, just a forEach. I would replace it with a for loop instead.

File: lib/ticks-to-tree.js [hunk 15]
```diff
@@ -99,71 +113,191 @@ function ticksToTree (ticks, options = {}) {
...
+      trees[tree].children.map((frame, index) => {
```
Comment (mcollina) [inline lib/ticks-to-tree.js]: same, this can be a for loop

File: lib/ticks-to-tree.js [hunk 16]
```diff
@@ -99,71 +113,191 @@ function ticksToTree (ticks, options = {}) {
...
+        renameFrame(options, frame, loadedSourceMaps)
+      })
     })
   }
+  Object.keys(loadedSourceMaps).forEach((file) => {
+    if (loadedSourceMaps[file] !== false) {
+      loadedSourceMaps[file].destroy()
+    }
+  })
```
Comment (mcollina) [inline lib/ticks-to-tree.js]: Use a for loop

File: lib/ticks-to-tree.js [hunk 17]
```diff
@@ -99,71 +113,191 @@ function ticksToTree (ticks, options = {}) {
...
+    if (typeof sourceMaps === 'object') {
+      if (sourceMaps[basename(fileName)]) {
+        sourceMapFiles.unshift(sourceMaps[basename(fileName)])
+      }
+      if (sourceMaps[fileName]) {
+        sourceMapFiles.unshift(sourceMaps[fileName])
+      }
+    }
+
+    if (access(fileName)) {
```
Comment (mcollina) [inline lib/ticks-to-tree.js]: are you sure this is sync?

File: lib/ticks-to-tree.js (hunk 9 above)
Comment (mcollina) [reply lib/ticks-to-tree.js]: Is it actually safe?

File: lib/ticks-to-tree.js (hunk 17 above)
Comment (DylanC) [reply lib/ticks-to-tree.js]: Oh its not actually. Updated now!

**Comment by DylanC** [comment]
> Does this include some unit tests for the transpiled case?

This will need to include some unit tests. An example exists but not a unit test.

**Comment by mcollina** [comment]
CI seems to be failing

**Comment by DylanC** [comment]
@mcollina Yep.

Its failing on this:
`not ok 2 - ENOENT: no such file or directory, access '[eval]'`

Line 167:
`if (accessSync(fileName)) {`

I'm looking into the flow of this feature at the moment. Something is not right here with the source map code.

**Pull request closed without merging**